activerecord/lib/active_record/scoping/named.rb

```diff
@@ -183,6 +204,23 @@ def scope(name, body, &block)
               scope
             end
           end
+
+          if predicate
+            method_name = if predicate.is_a?(Symbol) || predicate.is_a?(String)
+              predicate.to_sym
+            elsif predicate == true
+              :"within_#{name}"
+            else
+              raise ArgumentError,
+                "Invalid value for 'predicate'. Expected a Symbol, " \
+                "a String, or true, but received #{predicate.inspect}"
+            end
+
+            define_method("#{method_name}?") do
+              self.class.send(name).exists?(id)
```

Comment:
(From https://x.com/chaadow/status/1738292824846500134?s=20 )

>Unfortunately, this will create an N+1 issue. The difference with enum is that the instance method defined by enum does not perform a SQL query.

> What would be nice is if there is a way to transform the conditions inside the scope into ruby to avoid performing the SQL query.

As stated by @kaspth there is this [alternative ruby gem](https://github.com/joeldrapper/action_state) that can handle some, but not all, scope attributes, in ruby

Reply:
I think this might still risk differences between values from the database and values from Ruby, but could this implementation utilize [ActiveRecord::Relation#where_values_hash](https://github.com/rails/rails/blob/b1d57fb8c87d771d9a74ac31cd9969e6806b03b2/activerecord/lib/active_record/relation.rb#L783-L789)?


```ruby
define_method("#{method_name}?") do
  where_values_hash = self.class.public_send(name).where_values_hash

  where_values_hash.all? { |name, value| public_send(name) == value }
end
```

Reply:
@seanpdoyle that won't work for `where.not` as `where_values_hash` only collects equality Arel nodes.

We may be able to do something at the `Arel::Nodes` level though. Here's a version I've been experimenting with:

```ruby
define_method "#{method_name}?" do
  self.class.public_send(scope).where_clause.send(:predicates).all? do |node|
    value, matcher = public_send(node.left.name), node.right

    case node
    when Arel::Nodes::Equality      then value == matcher.value
    when Arel::Nodes::NotEqual      then value != matcher.value
    when Arel::Nodes::HomogeneousIn
      values = matcher.map(&:value)
      node.type == :notin ? !values.include?(value) : values.include?(value)
    else
      raise "can't generate an automatic query for this where clause chain"
    end
  end
end
```

Which will recognize something like this `where(id: 1).where.not(id: [2,3])`. It won't typecast the `in` though so `where.not(id: ["1"])` won't become an integer.

I'm not sure how feasible this is but it was fun to play with.

Reply:
@seanpdoyle @kaspth thank you both for exploring these options. Just so I understand, those explorations return how the query is to be structured, but don't actually execute SQL, correct?

If so, I can understand how this would work with a simple `where` clause with a **static** value, but would it work with something more complex and dynamic? Here's an example from a client project:

```ruby
scope :paid, -> {
  left_joins(:paid_payments, :paid_intake_co_payment)
    .group("bills.id")
    .having(<<-SQL)
      bills.patient_responsibility = 0 OR
      (
        (COALESCE(SUM(payments.amount), 0) * 100)::integer +
        (COALESCE(SUM(paid_intake_co_payments_bills.amount), 0) * 100)::integer
        >= bills.patient_responsibility AND bills.patient_responsibility > 0
      )
    SQL
}
```

Reply:
@kaspth 

> Which will recognize something like this where(id: 1).where.not(id: [2,3]). It won't typecast the in though so where.not(id: ["1"]) won't become an integer.

Is it possible to cast `matcher.value` based on `node.right.type`? 

```ruby
node.right.type

# => #<ActiveRecord::ConnectionAdapters::SQLite3Adapter::SQLite3Integer>